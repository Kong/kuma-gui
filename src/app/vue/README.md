---
section: Packages
---
# vue

Automatic `vue` initialization using our projects dependency injection layer.

