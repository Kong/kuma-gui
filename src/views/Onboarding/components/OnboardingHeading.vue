<template>
  <div class="relative">
    <h1 class="onboarding-title">
      {{ title }}
    </h1>
    <p
      v-if="description"
      class="text-center text-lg mt-3"
    >
      {{ description }}
    </p>
    <div
      v-if="showSkip"
      class="absolute right-0 top-0 mt-4 m-a flex items-center"
    >
      <KButton
        class="skip-button"
        appearance="btn-link"
        size="small"
        @click.native="skipOnboarding"
      >
        Skip Setup
      </KButton>
    </div>
  </div>
</template>

<script>
import { mapActions } from 'vuex'
export default {
  name: 'OnboardingHeading',
  props: {
    title: {
      type: String,
      required: true,
    },
    description: {
      type: String,
      default: '',
    },
    showSkip: {
      type: Boolean,
      default: true,
    },
  },
  methods: {
    ...mapActions('onboarding', ['completeOnboarding']),
    skipOnboarding() {
      this.completeOnboarding()
      this.$router.push({
        name: 'global-overview',
        params: {
          mesh: 'all',
        },
      })
    },
  },
}
</script>

<style lang="scss" scoped>
.onboarding-title {
  @apply text-center text-4xl font-bold;

  background: -webkit-linear-gradient(#260d50, #822dc5);
  -webkit-background-clip: text;
  -webkit-text-fill-color: transparent;
}

.skip-button {
  @apply font-medium;

  --KButtonBtnLink: rgba(38, 13, 80, 0.3);
}
</style>
