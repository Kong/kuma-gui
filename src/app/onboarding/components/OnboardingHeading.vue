<template>
  <div class="onboarding-heading">
    <h1 class="onboarding-title">
      <slot name="title" />
    </h1>

    <div
      v-if="slots.description"
      class="onboarding-description"
    >
      <slot name="description" />
    </div>
  </div>
</template>

<script lang="ts" setup>
import { useSlots } from 'vue'

const slots = useSlots()
</script>

<style lang="scss" scoped>
.onboarding-heading {
  position: relative;
}

.onboarding-title {
  padding-top: $kui-space-20;
  padding-bottom: $kui-space-20;
  text-align: center;
  font-size: 2.25rem;
  line-height: 2.5rem;
  font-weight: $kui-font-weight-semibold;
  color: transparent;
  background-image: linear-gradient(to right, var(--onboarding-heading-1), var(--onboarding-heading-2));
  -webkit-background-clip: text;
  background-clip: text;
}

.onboarding-description {
  margin-top: $kui-space-50;
  text-align: center;
  font-size: 1.125rem;
  line-height: 1.75rem;
}
</style>
