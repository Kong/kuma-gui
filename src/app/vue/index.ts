import { createApp } from 'vue'
import {
  createRouter,
  createWebHistory,
} from 'vue-router'

import type { ServiceDefinition } from '@/services/utils'
import { token, createInjections } from '@/services/utils'
import type { Component } from 'vue'
import type { Router, RouteRecordRaw, NavigationGuard } from 'vue-router'

type Token = ReturnType<typeof token>

type VueApp = ReturnType<typeof createApp>
export type PluginDefinition = Parameters<VueApp['use']>
export type ComponentDefinition = [string, Component]

const $ = {
  app: token<(App: Component) => Promise<VueApp>>('vue.app'),
  router: token<Router>('vue.router'),

  components: token('vue.components'),
  plugins: token<PluginDefinition[]>('vue.plugins'),
  routes: token<RouteRecordRaw[]>('vue.routes'),
  routesLabel: token<RouteRecordRaw[]>('vue.routes.label'),
  navigationGuards: token<NavigationGuard[]>('vue.routes.navigation.guards'),
}
export const services = (app: Record<string, Token>): ServiceDefinition[] => {
  return [
    [$.app, {
      service: (
        components: ComponentDefinition[],
        plugins: PluginDefinition[],
      ) => {
        return async (App: Component) => {
          const app = createApp(App)

          plugins.forEach(([...args]) => {
            app.use(...args)
          })

          components.forEach(([name, component]: [string, Component]) => {
            app.component(name, component)
          })

          return app
        }
      },
      arguments: [
        $.components,
        $.plugins,
      ],
    }],
    [$.router, {
      service: (
        env: (str: 'KUMA_BASE_PATH') => string,
        routes: RouteRecordRaw[],
        guards: NavigationGuard[],
      ) => {
        const router = createRouter({
          history: createWebHistory(env('KUMA_BASE_PATH')),
          routes,
        })

        guards.forEach((item) => {
          if (typeof item === 'function') {
            router.beforeEach(item)
          }
        })

        return router
      },
      arguments: [
        app.env,
        $.routes,
        $.navigationGuards,
      ],
    }],

    [token('vue.plugins'), {
      service: (router) => {
        return [
          [router],
        ]
      },
      arguments: [
        $.router,
      ],
      labels: [
        $.plugins,
      ],
    }],

    // TODO(jc): Label decoration. Make it so we can decorate labels
    // Temporarily turn the routesLabel label into routes for decoration
    // purposes
    [$.routes, {
      service: (routes: RouteRecordRaw[]) => routes,
      arguments: [
        $.routesLabel,
      ],
    }],
    //

    // TODO(jc): Make it so we don't need to provide a [] empty default for
    // labels
    [token('application.routes.navigation.guards'), {
      service: () => {
        return []
      },
      labels: [
        $.navigationGuards,
      ],
    }],
  ]
}

export const TOKENS = $
export const [
  useRouter,
] = createInjections(
  TOKENS.router,
)
