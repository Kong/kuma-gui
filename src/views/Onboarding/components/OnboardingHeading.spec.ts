import { render } from '@testing-library/vue'

import OnboardingHeading from './OnboardingHeading.vue'

describe('OnboardingHeading.vue', () => {
  it('renders snapshot', () => {
    const { container } = render(OnboardingHeading, {
      props: {
        title: 'title',
        description: 'description',
      },
    })

    expect(container).toMatchSnapshot()
  })
})
