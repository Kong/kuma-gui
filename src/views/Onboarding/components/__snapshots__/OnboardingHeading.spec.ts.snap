// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OnboardingHeading.vue renders snapshot 1`] = `
<div>
  <div
    class="relative"
  >
    <h1
      class="onboarding-title"
    >
      
    title
  
    </h1>
     
    <p
      class="text-center text-lg mt-3"
    >
      
    description
  
    </p>
     
    <div
      class="absolute right-0 top-0 mt-4 m-a flex items-center"
    >
      <button
        class="k-button skip-button small btn-link"
        data-v-38fa8ecf=""
        type="button"
      >
        
      Skip Setup
    
      </button>
    </div>
  </div>
</div>
`;
